proxy/l4r/L4RProxyServerMain.cc

```diff
@@ -0,0 +1,408 @@
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include "ts/ink_config.h"
+#include "P_Net.h"
+#include "Main.h"
+//#include "HttpConfig.h"
+//#include "HttpSessionAccept.h"
+//#include "ReverseProxy.h"
+//#include "HttpSessionManager.h"
+//#include "HttpUpdateSM.h"
+#ifdef USE_HTTP_DEBUG_LISTS
+//#include "Http1ClientSession.h"
+#endif
+//#include "HttpPages.h"
+//#include "HttpTunnel.h"
+#include "ts/Tokenizer.h"
+#include "P_SSLNextProtocolAccept.h"
+#include "ProtocolProbeSessionAccept.h"
+//#include "http2/Http2SessionAccept.h"
+//#include "HttpConnectionCount.h"
+#include "L4RProxyServerMain.h"
+#include "L4RSessionAccept.h"
+
+#include <vector>
+
+//HttpSessionAccept *plugin_http_accept             = nullptr;
+//HttpSessionAccept *plugin_http_transparent_accept = nullptr;
+
+//static SLL<SSLNextProtocolAccept> ssl_plugin_acceptors;
+//static Ptr<ProxyMutex> ssl_plugin_mutex;
+
+//std::mutex proxyServerMutex;
+//std::condition_variable proxyServerCheck;
+//bool et_net_threads_ready = false;
+
+extern int num_of_net_threads;
+extern int num_accept_threads;
+
+/// Global BufferWriter format name functions.
+namespace
+{
+void
+TS_bwf_thread(ts::BufferWriter &w, ts::BWFSpec const &spec)
+{
+  bwformat(w, spec, this_thread());
+}
+void
+TS_bwf_ethread(ts::BufferWriter &w, ts::BWFSpec const &spec)
+{
+  bwformat(w, spec, this_ethread());
+}
+} // namespace
+
+// File / process scope initializations
+static bool L4R_SERVER_INITIALIZED __attribute__((unused)) = []() -> bool {
+  ts::bwf_register_global("ts-thread", &TS_bwf_thread);
+  ts::bwf_register_global("ts-ethread", &TS_bwf_ethread);
+  return true;
+}();
+
+#if 0
+bool
+ssl_register_protocol(const char *protocol, Continuation *contp)
+{
+  SCOPED_MUTEX_LOCK(lock, ssl_plugin_mutex, this_ethread());
+
+  for (SSLNextProtocolAccept *ssl = ssl_plugin_acceptors.head; ssl; ssl = ssl_plugin_acceptors.next(ssl)) {
+    if (!ssl->registerEndpoint(protocol, contp)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool
+ssl_unregister_protocol(const char *protocol, Continuation *contp)
+{
+  SCOPED_MUTEX_LOCK(lock, ssl_plugin_mutex, this_ethread());
+
+  for (SSLNextProtocolAccept *ssl = ssl_plugin_acceptors.head; ssl; ssl = ssl_plugin_acceptors.next(ssl)) {
+    // Ignore possible failure because we want to try to unregister
+    // from all SSL ports.
+    ssl->unregisterEndpoint(protocol, contp);
+  }
+
+  return true;
+}
+#endif
+
+/////////////////////////////////////////////////////////////////
+//
+//  main()
+//
+/////////////////////////////////////////////////////////////////
+
+/** Data about an acceptor.
+
+    This is used to separate setting up the proxy ports and
+    starting to accept on them.
+
+*/
+struct L4RProxyAcceptor {
```

Comment:
This should go in to "iocore/net" along with the other accept logic.

Reply:
Although, really, you shouldn't need a separate Acceptor at all. For example, if the inbound connection is TLS the determination to L4 proxy it can't be made until after the Client Hello packet has been recieved, which is long after all the accept logic.

Reply:
The Layer 4 proxying must be able to run on the same ports as other traffic, it can't require dedicate proxy ports or executables.

Reply:
Understood. But I thought for non SSL in, non SSL out case we will have to make the decision based on the port number.